components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/CacheManager.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.mediator.cache;
+
+import com.google.common.cache.LoadingCache;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * There would be two instances of the cache mediator in a single mediation flow. Hence it must be possible for the
+ * cache created in one instance to be reused in the next. This CacheManager enables this feature with static methods.
+ */
+class CacheManager {
```

Comment:
Why can't the CacheMediator use the standard Java Caching API? What made you write a separate CacheManager?

Have you tested how this works in a multi-tenant scenario? If so, what was the behavior?

Should we not maintain separate cachemanagers for each tenant? 
Did the cache got cleaned up during tenant unload in the current design? 
What tests were carried out to make sure there are no memory leaks?

Reply:
Using static methods, and assuming singleton behavior is quite dangerous in here IMO.

Reply:
The javax cache do not support certain features like the maximum size of cache. To get the improved features of caching we use guava caching.

Yes we have tested it in a multi-tenant scenario. There is a separate cache for each `CacheMediator` instance which is identified by an id. To manage all these caches a `CacheManager` has been used. Whenever the artifact(sequence, proxy etc.) are undeployed the cache related to that artifact is removed from the cache manager. Because of separate caches for each artifact we do not have a separate cache manager for each tenant. 

The cache gets cleaned up whenever an artifact is undeployed and hence whenever the tenant is unloadee.

The main reason for using to guava caching is to avoid memory leaks. It is possible to specify the maxMessageSize and the maxSize of the cache and timeout for each cache entry and hence cache is evicted in the appropriate manner.

Reply:
We understand that using static methods are dangerous and will fix and update soon.

Reply:
Adding @afkham to get his opinion.

> The javax cache do not support certain features like the maximum size of cache. To get the improved features of caching we use guava caching.

@riyafa your observation that JCache API does not define a capacity is correct. The spec has let the implementations to decide what upper limits to use. Rationale is that capacity could mean anything - It may be the no. of entries, heapsize taken by the cache etc. 

Our JCache implementation does have support for max capacity of the cache. See [1]. It is not good to bring in guava caching when we already have a caching implementation. Could you refer me to the discussion thread please?

> Yes we have tested it in a multi-tenant scenario. There is a separate cache for each CacheMediator instance which is identified by an id. To manage all these caches a CacheManager has been used. Whenever the artifact(sequence, proxy etc.) are undeployed the cache related to that artifact is removed from the cache manager. Because of separate caches for each artifact we do not have a separate cache manager for each tenant.

> We used CacheManager to globally manage caches for each instance of the CacheMediator. We understand that using static methods are dangerous and this has been fixed in #890 

Cool. 
IMO, what you need here is a Cache though, not a CacheManager. That Cache may contain set of cache entries which may contain the http responses. Please read the [jcache Interface Summary](https://static.javadoc.io/javax.cache/cache-api/1.0.0/javax/cache/package-summary.html) for more details. 

Further, this cache instance can be local to the CacheMediator because it is actually part of that. Then, the advantage is that the Cache instance will get garbage collected when the CacheMediator instance dies. So, IMHO, we can improve this design. :)

[1] https://github.com/wso2/carbon-kernel/blob/4.4.x/core/javax.cache/src/main/java/org/wso2/carbon/caching/impl/CacheImpl.java#L838

Thanks.